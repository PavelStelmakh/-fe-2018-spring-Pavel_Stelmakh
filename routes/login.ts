import { Router, Response, Request } from 'express';
const router = Router();
import { getUserByToken } from '../shared/getUserByToken';
import { getToken } from '../shared/getToken';
import { UserAuth } from '../models/UserAuth';
import { User } from '../models/User';
const users = require('../models/Users');

router.get('', function (req: Request, res: Response) {
    const {user} = getUserByToken(req);
    
    if (user) {
        const data: {id: number, role: string} = {id: user['id'], role: user['role']};
        res.status(200).send(data);
    } else {
        res.status(401);
    }
    res.end();
    
});

router.get('/find/:name/:id', function (req: Request, res: Response) {
    const name: string = req.params['name'];
    const id: number = +req.params['id'];

    setTimeout(() => {
        if (id !== 0 && users.user(id) && users.user(id).name === name) {
            res.status(200);
        } else if (!users.checkExistName(name)) {
            res.status(200);
        } else {
            res.status(400);
        }
        res.end();
    }, 3000);
});

router.get('/logout', function (req: Request, res: Response) {
    const data = getUserByToken(req);
    
    if (data.user) {
        res.clearCookie('auth_token');
        res.status(200);
    } else {
        res.status(401);
    }
    res.end();
});

router.post('', function (req: Request, res: Response) {
    if (req.body) {
        const login = req.body.login;
        const password = req.body.password;

        if (login && password) {
            const user: UserAuth = users.checkLogin({login, password});
            if (user.isAuthenticated === true) {
                const secret = req.app.get('secret');
                const token = getToken(user.user, secret);
                const role: string = (user.user as User).role;

                res.cookie('auth_token', token);                
                res.status(200).send({role});
            } else {
                res.status(401);
            }
        } else {
            res.status(401);
        }
    } else {
        res.status(400);
    }
    
    res.end();
});

router.post('/recovery', function (req: Request, res: Response) {
    if (req.body) {
        const login = req.body.login;
        const birthday = req.body.birthday;
        const password = req.body.password;

        if (users.recoveryLogin({login, birthday, password})) {
            res.status(200);
        } else {
            res.status(401);
        }
    } else {
        res.status(400);
    }
    
    res.end();
});

module.exports = router;