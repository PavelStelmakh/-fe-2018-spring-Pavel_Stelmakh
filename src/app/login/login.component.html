<form [formGroup]="loginForm" (ngSubmit)="onSubmit()">
  <div [class.messages]="true">
    <div *ngIf="wrongData && login.pristine">{{'login.wrongData' | translate}}</div>
    <div *ngIf="login.dirty && login.errors?.checkExistName">{{ 'login.notFound' | translate }}</div>
    <div class="loading" *ngIf="login.dirty && login.status === 'PENDING'">loading...</div>
  </div>
  <input type="text" placeholder="login" formControlName="login">
  <input type="password" placeholder="password" formControlName="password">
  <div [class.singIn]="true">
    <input type="submit" [disabled]="loginForm.pristine || loginForm.status !== 'VALID'" value="{{'login.singnIn' | translate}}">
    <a routerLink="/login/recovery">{{'login.fogotPassword' | translate}}</a>
  </div>
</form>
