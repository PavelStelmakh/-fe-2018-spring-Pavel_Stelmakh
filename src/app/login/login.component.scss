@keyframes loginAnimation {
    from {
        top: 0;
        background-color: rgba(0, 0, 0, 0);
        box-shadow: 0 0 0px 0px rgba(0, 0, 0, 0);
    }
    to {
        top: 20vh;
        background-color: rgba(0, 0, 0, 0.7);
        box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.5);
    }
}

@keyframes messageAnimation {
    from {
        top: -40px;
        background-color: rgba(250, 86, 85, 0);
        box-shadow: 0 0 0px 0px rgba(0, 0, 0, 0);
    }
    to {
        top: 0;
        background-color: rgba(250, 86, 85, 1);
        box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.5);
    }
}

:host {    
    text-align: center;
    display: block;
    width: 500px;
    background-color: rgba(0, 0, 0, 0.7);
    position: absolute;
    top: 20vh;
    left: calc(50vw - 250px);
    box-shadow: 0 0 10px 3px rgba(0,0,0,0.5);
    animation: loginAnimation 1s;

    form {
        padding: 10px 0;
    }

}

.singIn {
    margin-top: 10px;
}

a {    
    color: #ccc;
    margin-left: 10px;
}

.messages {
    position: absolute;
    top: -13vh;
    height: 40px;
    line-height: 40px;
    width: 500px;
    color: #ccc;

    div {
        position: relative;
        top: 0;
        background-color: rgba(250, 86, 85, 1);
        box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.5);
        animation:  messageAnimation 1s;
    }

    .loading {
        background-color: rgb(250, 86, 85);
    }

}