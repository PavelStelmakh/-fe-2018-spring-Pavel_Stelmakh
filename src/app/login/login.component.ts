import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { AuthService } from '../auth/auth.service';
import { Router } from '@angular/router';
import { SignIn } from 'models/SignIn';
import { checkExistNameValidator } from '../validators/checkExistName-validator';
import { UsersService } from '../users.service';

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.scss']
})
export class LoginComponent implements OnInit {
  loginForm: FormGroup;
  wrongData: boolean;

  constructor(
    private builder: FormBuilder, 
    private auth: AuthService,
    private router: Router,
    private usersService: UsersService
    ) {}

  ngOnInit() {
    this.wrongData = false;
    this.loginForm = this.builder.group({
      login: ['', Validators.required, checkExistNameValidator(this.usersService, true)],
      password: ['', Validators.required]
    });
  }

  get login() {
    return this.loginForm.get('login');
  }

  get password() {
    return this.loginForm.get('password');
  }

  onSubmit() {
    const data: SignIn = {
      login: this.login.value,
      password: this.password.value
    };
    this.auth.signIn(data).subscribe(result => {
      if (result.status === 200) {        
        this.router.navigate(['/details']);
      }
    },
    error => {
      this.loginForm.reset();
      this.wrongData = true;
    });
    
  }

}
