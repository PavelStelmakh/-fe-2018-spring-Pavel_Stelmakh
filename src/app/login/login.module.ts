import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LoginComponent } from './login.component';
import { Routes, RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';
import { RecoveryLoginComponent } from '../recovery-login/recovery-login.component';
import { AuthGuard } from '../auth/auth.guard';
import { HttpClientModule, HttpClient } from '@angular/common/http';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { HttpLoaderFactory } from '../shared/HttpLoaderFactory';

const route: Routes = [
  {
    path: 'login',
    component: LoginComponent,
    canActivate: [AuthGuard]
  },
  {
    path: 'login/recovery',
    component: RecoveryLoginComponent,
    canActivate: [AuthGuard]
  }
]

@NgModule({
  imports: [
    CommonModule, 
    RouterModule.forChild(route),
    ReactiveFormsModule,
    HttpClientModule,
    TranslateModule.forRoot({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
})
  ],
  declarations: [LoginComponent, RecoveryLoginComponent],
  exports: [LoginComponent]
})
export class LoginModule { }
