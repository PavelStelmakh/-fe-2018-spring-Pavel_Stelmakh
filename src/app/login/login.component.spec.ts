import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from "@angular/router/testing";
import { HttpClientModule } from '@angular/common/http';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { LoginComponent } from './login.component';
import { BASE_URL } from '../url.service';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';

const translations = {
  "login": {
    "wrongData": "wrong data",
    "notFound": "such login is not found",
    "fogotPassword": "forgot password?",
    "singnIn": "Sign in"
}
};

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('LoginComponent', () => {
  let fixture: ComponentFixture<LoginComponent>;
  let translate: TranslateService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        ReactiveFormsModule,
        HttpClientModule,
        HttpClientTestingModule,
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        })
      ],
      declarations: [
        LoginComponent
      ],
      providers: [
        {provide: BASE_URL, useValue: 'http://localhost:3000'}
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(LoginComponent);
    translate = TestBed.get(TranslateService);
    translate.use('en');
  });

  it('renders markup to snapshot', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

});
