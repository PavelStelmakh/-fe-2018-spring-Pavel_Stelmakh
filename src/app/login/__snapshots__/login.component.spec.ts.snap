// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent renders markup to snapshot 1`] = `
<app-login
  auth={[Function AuthService]}
  builder={[Function FormBuilder]}
  loginForm={[Function FormGroup]}
  router={[Function Router]}
  usersService={[Function UsersService]}
  wrongData="false"
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <div
      class="messages"
    >
      
      
      
    </div>
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="login"
      ng-reflect-name="login"
      placeholder="login"
      type="text"
    />
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="password"
      ng-reflect-name="password"
      placeholder="password"
      type="password"
    />
    <div
      class="singIn"
    >
      <input
        disabled=""
        type="submit"
        value="Sign in"
      />
      <a
        href="/login/recovery"
        ng-reflect-router-link="/login/recovery"
        routerlink="/login/recovery"
      >
        forgot password?
      </a>
    </div>
  </form>
</app-login>
`;
